### PR TITLE
gooddata writer v3 via config rows

### PR DESCRIPTION
**Pull request by kacurez**
Fixes https://github.com/keboola/kbc-ui/issues/1574

issue implementacie: https://github.com/keboola/kbc-ui/issues/1574
issue dizajnu: https://github.com/keboola/indigo-ui/issues/40 
jackov prototyp: https://invis.io/W3FV2MIMV59
komponenta: https://github.com/keboola/gooddata-writer-v3

### TLDR;
- `update config rows` pomocou predlohy Jackovho navrhu ui gooddata writeru  https://github.com/keboola/kbc-ui/tree/kacurez-goodata-writer-v3/src/scripts/modules/configurations 
- `POC gooddata writer v3` postaveny nad upravou config rows https://github.com/keboola/kbc-ui/tree/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3
- `compatibility update ex http a s3` tj existujucich ui kbc komponent nad config rows 
    - ex-aws-s3 (https://github.com/keboola/kbc-ui/pull/1589/commits/efbf246cd76619becba965de617e54fa46c2ad46) 
    - ex-http (https://github.com/keboola/kbc-ui/pull/1589/commits/42bb8f605031222a516f3d5ac2db1c47a8f3dabc)

### Vizualne zmeny oproti sucasnemu ui
~Je vlastne len jedna a to presun save buttons do headru stranky detailu row~ + samozrejme gooddata writer ui ale ta komponenta je este skryta a bude sa na nej dalej pracovat.
<details>
<summary>~Priklad detailu stranky ex s3 s buttons v headru stranky~</summary>

</details>

### Co dalej po merge do masteru
Nasledujuce prace bude mozne robit separatne a nezavisle na sebe s minimalnou upravou funcknosti config rows:
- dokoncit Goodata Writer ui, to by som rozdelil do 2 etap:
    - 1. plne funkcny prototyp
    - 2. postylovanie + integracia s indigo-ui
- cez config rows napisat dalsiu komponentu/writer napr google big query
- stylovanie config rows - layout a spolocne ui komponenty ako buttony, wrapper komponenty(napr collpasible section, tabulka config rows, table columns editor)

#### Sections

Custom content Index a Row page je definovany polom objektov tzv sections. Section je "znamy adapter" objekt s properties `render, onLoad, onSave, onCreate, isComplete`(zmizol title) + je mozne definovat onSave, onLoad a onCreate na globalnej urovni danej stranky, ak tam niesu tak sa pouziju "neutralne" defaulty. Pri nacitavani stranky sa konfig rozparsuje cez globalnu onLoad fci a potom do jednotlivych sekci cez ich onLoad fce. Podobne onSave vytvori vysledny konfig  zmergovani vysledkov onCreate zo sekci a zatym globalny onCreate. 

<details>
<summary>
Gooddata writer row sections z pohladu renderu v ui a ich definicia v kode
</summary>
<a href="https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3/routes.js#L39-L56">
https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3/routes.js#L39-L56
</a>
<span>

</span>
</details>

#### Nove react properties rendru sekce - onSave a isComplete
Do renderu sekce okrem `onChange, disabled a value` pribudli react properties:
- `onSave` cez ktoru render komponente umoznuje ulozit do konfigu svoju value. Zaroven to nijak neovplyvnuje stav ostatnych rozeditovanych sekci. Vyuziva sa to napr v gooddata writery na definiciu date dimension ktore sa podla Jackovho navrhu vytvaraju cez vlastny modal dialog tu:
https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3/react/components/DimensionsSection.jsx#L14
Dalsim vyuzitim moze byt napr oauth sekce..
- `isComplete` ktora sa vyuziva na auto collapse sekce v collapsible section komponente napr tu:
https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/configurations/utils/createCollapsibleSection.jsx#L33

#### Utility na preddefinovanie seckci

Koncept utilit ktore na zaklade predpisu vygeneruju cely objekt pre definiciu sekce alebo jej cast, zatial su 2:
- 1. auto collapsible section(accordion) s opt-in save buttons(https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/configurations/utils/createCollapsibleSection.jsx) vyuzitie napr v https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3/routes.js#L48 alebo https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/ex-aws-s3/routes.js#L23

- 2. Storage Table Columns Editor - casto vyuzivana komponenta vo writeroch na definiciu datovych typov alebo ineho info pre kazdy zapisovany stlpec tabulky(https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/configurations/utils/createColumnsEditorSection.js). 
Priklad pouzita pre gooddata writer: https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3/routes.js#L55
Vstupom su definicie stlpcov(render) a ich nacitanie/uloznie z konfigu. Nacitanie zoznamu stlpcov zo storage a ich data preview sa generuje automaticky tj neriesi sa v jeho definicii(vid https://github.com/keboola/kbc-ui/blob/kacurez-goodata-writer-v3/src/scripts/modules/gooddata-writer-v3/adapters/columnsEditorDefinition.js)
 
<details>
<summary>GoodData writer columns editor render vs code</summary>

</details>

### CODE REVIEW

File: src/scripts/modules/gooddata-writer-v3/react/components/DimensionsSection.jsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            saveLabel="Create"
+            onCancel={this.closeModal}
+            onSave={this.handleCreate}/>
+        </Modal.Footer>
+      </Modal>
+    );
+  },
+
+  isValid() {
+    return true;
```
Comment (ujovlado) [inline src/scripts/modules/gooddata-writer-v3/react/components/DimensionsSection.jsx]: toto sa dufam planuje nejako rozsirovat :) ak nie, vyhod to pls

**Comment by ujovlado** [comment]
Este som si vsimol ze v gooddata writeri zmazanie riadku z jeho detailu nefunguje, hodi ma to na index, riadok tam je, po kliknuti na detail error.

**Comment by ujovlado** [comment]
Inac sa mi tie komponenty celkom pacia - tie co su potom pouzite v konkretnych komponentach.

Jedine co ma tam trosku zacina hnevat je hromada kodu v samotnom `configurations` - aby to tam nezacalo byt potom neprehladne. Uz teraz sa v tom orientuje tazko.

Dalej to reviewovat asi nebudem, nechal by som to na @ondrejhlavacek - s nim ste to riesili.

Poslime to ale cim skor do produkcie, nech sa z toho nestane 5k riadkova vec ;)

**Comment by ondrejhlavacek** [comment]
@ujovlado k review se dostanu během tohodle týdne, nejpozději začátkem příštího, už mám sám kudlu na krku. nic tam, pokud vím, přidělávat dalšího nebudem, mám sice nějaký komponenty, který v tom chci napsat, ale to už bude novej čistej pull request, kterej bude navazovat. omlouvám se za delay!

File: src/scripts/modules/gooddata-writer-v3/react/components/NewDimensionForm.jsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+              disabled={disabled}
+              onChange={e => this.handleChange({identifier: e.target.value})}
+              value={value.identifier}
+            />
+          </Col>
+        </FormGroup>
+        <FormGroup>
+          <Col componentClass={ControlLabel} sm={3}>Template</Col>
+          <Col sm={9}>
+            <RadioGroup
```
[the remark this answers is not in the document]
Comment (kacurez) [reply src/scripts/modules/gooddata-writer-v3/react/components/NewDimensionForm.jsx]: myslis `RadioGroup`? neni problem, som nevedel ze je obsolete :)

**Comment by kacurez** [comment]
@ujovlado diky za review. Tie poznamky ohladom samotneho gooddata-writeru poriesim v dalsom PR, tu som to naprgal nahrubo a hlavne to sluzi na overenie konceptu "config-rows".
Inac suhlas stym mnozstvom kodu v `configurations`, tiez ma to stve, a chcem to postupne refactorovat.

@ondrejhlavacek mozme si to spolu prejst a nabootovat sa ak by si chcel..

**Comment by kacurez** [comment]
@ujovlado este sa vratim k tomu zmazaniu riadku co pises vyssie ze neide - skusam to a nedari sa mi to nasimulovat tak neviem, ale asi nato este narazim(e).

**Comment by ondrejhlavacek** [comment]
@kacurez tak jo, dneska po obědě?

**Comment by kacurez** [comment]
@ondrejhlavacek ok mozem

File: src/scripts/modules/ex-aws-s3/routes.js [hunk 3]
```diff
@@ -12,29 +12,30 @@ import {
...
   componentId: 'keboola.ex-aws-s3',
   componentType: 'extractor',
-  credentials: {
-    show: true,
-    detail: {
-      title: 'AWS Credentials',
-      render: CredentialsForm,
+  index: {
+    sections: [{
+      render: createCollapsibleSection(() => <span>AWS Credentials</span>,
```
Comment (ondrejhlavacek) [inline src/scripts/modules/ex-aws-s3/routes.js]:
```
{
    render: TitleSection
},
{
    render: {
        component: TitleSection
    }
},
{
    render: {
        type: constants.COLLAPSIBLE,
        title: (value) => {...},
        component: LoadTypeSection,
        collapsed: (value) => {...}
    }
}

if (render is React...)
```

File: src/scripts/modules/ex-aws-s3/routes.js [hunk 4]
```diff
@@ -12,29 +12,30 @@ import {
...
+                                       CredentialsForm,
+                                       {includeSaveButtons: true}),
       onSave: credentialsCreateConfiguration,
       onLoad: credentialsParseConfiguration,
       isComplete: credentialsIsComplete
```
Comment (ondrejhlavacek) [inline src/scripts/modules/ex-aws-s3/routes.js]: `isComplete` možná přesunout do `render.collapsed`/`render.isCollapsed`

**Comment by ondrejhlavacek** [comment]
what's next:

- [ ] zapracovat současný připomínky (kacurez)
- [ ] přesunout save tlačítko zpět, přesun doprava nahoru pak pro všechny komponenty naráz v separátním PR
- [ ] smazat dummy code
- [ ] proklikat S3 extraktor a HTTP extraktor (najlos)
- [ ] mergnout
- [ ] storage api writer (najlos)
- [ ] gooddata writer (kacurez)
- [ ] refactoring
- [ ] bigquery writer / salesforce extraktor (najlos)

File: src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      rowId,
+      parseBySectionsFn,
+      createBySectionsFn
+    ) || !isParsableConfiguration(
+      Store.getConfiguration(componentId, configurationId, rowId),
+      parseBySectionsFn,
+      createBySectionsFn
+    );
+
+    const isJsonConfigurationSaving = isJsonEditorOpen && Store.getPendingActions(componentId, configurationId, rowId).has('save-json');
```
Comment (ondrejhlavacek) [inline src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx]: jakto, že tohle prošlo linterem? má to 137 znaků a nedá se to na githubu přečíst. všechny dlouhý řádky by to chtělo zalomit.

File: src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      <div className="text-right">
+        <SaveButtons
+          disabled={!this.state.isComplete}
+          isSaving={this.state.isSaving}
+          isChanged={this.state.isChanged}
+          onSave={this.handleSave}
+          onReset={this.handleReset}
+          showModal={this.state.showModal}
+          modalTitle="Save Parameters"
+          modalBody={(<div>The changes in the configuration are not compatible with the original visual form. Saving this configuration will disable the visual representation of the whole configuration and you will be able to edit the configuration in JSON editor only.</div>)}
```
Comment (ondrejhlavacek) [inline src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx]: zalomit pls

File: src/scripts/modules/configurations/react/components/StorageTableColumnsEditor.jsx [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            </th>
+          </tr>
+        </thead>
+        {this.renderBody()}
+      </table>
+    );
+  },
+
+  onChangeColumn(column) {
+    const newColumns = this.props.value.columns.map( c => c.id === column.id ? column : c);
```
Comment (ondrejhlavacek) [inline src/scripts/modules/configurations/react/components/StorageTableColumnsEditor.jsx]: nepojmenovávat proměnný jedním písmenkem prosím
a mezera za závorkou před c `( c =>` je legální?

File: src/scripts/modules/configurations/react/pages/Row.jsx [hunk 8]
```diff
@@ -285,22 +339,28 @@ export default React.createClass({
...
+      settings.getIn(['row', 'detail', 'onLoad'])
+    );
     const Configuration = this.state.settings.getIn(['row', 'detail', 'render']);
     return (<Configuration
-      onChange={function(diff) {
-        Actions.updateConfiguration(state.componentId, state.configurationId, state.rowId, configuration.merge(Immutable.fromJS(diff)));
-      }}
-      disabled={this.state.isSaving}
-      value={configuration.toJS()}
+              onChange={(diff) =>
```
Comment (ondrejhlavacek) [inline src/scripts/modules/configurations/react/pages/Row.jsx]: divnej indent?

File: src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx (hunk 5 above)
Comment (kacurez) [reply src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx]: tie rowhederbuttons sme sa dohodli ze dame prec - spat na index stranku takze to bude vypadat inak..

File: src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx (hunk 6 above)
Comment (kacurez) [reply src/scripts/modules/configurations/react/components/RowHeaderButtons.jsx]: detto ^^

File: src/scripts/modules/configurations/utils/createColumnsEditorSection.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+         columnsKey = 'columns', // property in config/localstate representing columns object/array
+         matchColumnKey = 'id', // used for matching columns objects with storage table columns by its name (columnName, columnObject) => true/false
+         columnsMappings = [], // array of object containing render and title property
+         isComplete = () => true, // is representation complete?
+         isColumnIgnored = column => column.get('type') === 'IGNORE'} = params; // if ignored then won't be saved to input mapping columns property of configuration object
+
+  return fromJS({
+    onSave(localState) {
+      const localStateColumnsToSave = localState.get(columnsKey)
+                                                .filter(c => !isColumnIgnored(c));
```
Comment (ondrejhlavacek) [inline src/scripts/modules/configurations/utils/createColumnsEditorSection.js]: divnej indent, neměl by tohle zase zachytit linter?

File: src/scripts/modules/configurations/utils/sections.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    sections: sectionsData
+  });
+
+  return createBySections(createFn, sectionsCreateFn, configurationBySections);
+}
+
+function isCompleteBySections(rootIsCompleteFn, sectionsIsCompleteFn, configBySections) {
+  const rootIsComplete = rootIsCompleteFn(configBySections.get('root'));
+  const sectionsIsComplete = configBySections.get('sections', List())
+                                             .reduce(
```
Comment (ondrejhlavacek) [inline src/scripts/modules/configurations/utils/sections.js]: divnej indent

File: src/scripts/modules/ex-http/routes.js [hunk 11]
```diff
@@ -13,27 +13,28 @@ import ConfigurationForm from './react/components/Configuration';
...
   componentType: 'extractor',
-  credentials: {
-    show: true,
-    detail: {
-      title: 'Base URL',
-      render: CredentialsForm,
+  index: {
+    sections: [{
+      render: createCollapsibleSection(() => <span>Base URL</span>,
+                                       CredentialsForm,
```
Comment (ondrejhlavacek) [inline src/scripts/modules/ex-http/routes.js]: tak tady ten indent už vůbec nechápu :-)

**Comment by ondrejhlavacek** [comment]
Do kódu jsem ještě hodil pár poznámek k indentování a dlouhejm řádkům, kdyžtak mě prosím uveďte na pravou míru, že to je v pohodě, jinak by se na to měl zavést linter.

File: src/scripts/modules/configurations/react/components/StorageTableColumnsEditor.jsx (hunk 7 above)
Comment (kacurez) [reply src/scripts/modules/configurations/react/components/StorageTableColumnsEditor.jsx]: asi to legalne je ked to prejde linterom :)
jedopismenkove premenne pouzivam ked je z kontextu jasne co znamenaju v tomto pripade je u toho map cez `columns` takze jeden item je logicky column, a potom teda c, premna je to citatelnejsie..

**Comment by ujovlado** [comment]
Az sa zbavime coffee, tak poladime aj ten linter. Indentoval by som teda 2 medzerami a pouzil sedliacky rozum alebo si to pastni do Prettiera, ked si nie si isty.

Priklanam sa ale k @ondrejhlavacek, urcite tak odsadenie nie je v ok a malo by to byt skor najako takto:

```js
  index: {
    sections: [{
      render: createCollapsibleSection(() => <span>Base URL</span>,
        CredentialsForm,
        {includeSaveButtons: true}),
      onSave: credentialsCreateConfiguration,
      onLoad: credentialsParseConfiguration,
      isComplete: credentialsIsComplete
    }]
  },
```
Alebo ked to pastnem do Prettiera s dlzkou riadku 120, a par dalsimi options, tak takto:

```js
  index: {
    sections: [
      {
        render: createCollapsibleSection(() => <span>Base URL</span>, CredentialsForm, { includeSaveButtons: true }),
        onSave: credentialsCreateConfiguration,
        onLoad: credentialsParseConfiguration,
        isComplete: credentialsIsComplete,
      },
    ],
  },
```

**Comment by ujovlado** [comment]
Spravim na to issue, toto by sme namali riesit  vobec, ma to byt automatizovane.

File: src/scripts/modules/configurations/react/components/StorageTableColumnsEditor.jsx (hunk 7 above)
Comment (kacurez) [reply src/scripts/modules/configurations/react/components/StorageTableColumnsEditor.jsx]: u named fci nikdy nepouzivam jednopismenove pomenovanie, no u anonymnych fci mi pride cim kratsie tym citatelnejsie, pretoze sa vzdy musis pozriet cez co sa vola map(this.props.value.columns). Kazdopadne mozem to vyskusat aj takto rozpisat, mozno to az tak vadit nebude :)

**Comment by kacurez** [comment]
ok budem pouzivat prettier ked si nebudem isty

**Comment by kacurez** [comment]
tie pripomienky na indent teda opravim

**Comment by kacurez** [comment]
nasetupoval som si prettier na onSave a preformatoval zdrojaky v okomentovanych moduloch

File: src/scripts/modules/ex-aws-s3/routes.js (hunk 3 above)
Comment (kacurez) [reply src/scripts/modules/ex-aws-s3/routes.js]: upravil som to podla toho ale zaroven narazil na problem ktory nepojde tak jednoducho odstranit. Tym ze sa React collabsible komponenta generuje pri renderi a nie pri inicializacii routes.js objektu(ako bolo predtym) tak sa vzdy vyresetuje state a tym isCollapsed stav. Riesenie spociva v komplexnej uprave `pages/Row.js` co zaberie cas, ale popravde sa mi to moc nepaci(stav isCollapsed by musel byt pod state Row.js). Zamna najlepsie by bolo nechat ako je teraz akurat upravit nazvoslovie a konfiguraciu, takze 
namiesto:
```
      render: {
        type: COLLABSIBLE_SECTION,
        title: () => <span>AWS Credentials</span>,
        component: CredentialsForm,
        options: {includeSaveButtons: true}
      },
```

revertnut https://github.com/keboola/kbc-ui/pull/1589/commits/7b1bee9b7ba5f29ec53d0d3c7a5fb0b94e744fbd a upravit na:

```
      render: CollabsibleSection({        
        title: () => <span>AWS Credentials</span>,
        component: CredentialsForm,
        options: {includeSaveButtons: true}
      }),
```

**Comment by ondrejhlavacek** [comment]
ref https://github.com/keboola/kbc-ui/issues/1629

**Comment by kacurez** [comment]
Upravil som to jednoduchsie, teda tak ako popisujem v https://github.com/keboola/kbc-ui/pull/1589#discussion_r187663191
Takze sa toto moze mergnut a k tomu problemu sa mozme vratit neskor. Zamna je to ready na merge, ak vadi naming tak to kludne opravte.(cc @ondrejhlavacek @ujovlado )

**Comment by ondrejhlavacek** [comment]
naming opravím a dám vědět

**Comment by ondrejhlavacek** [comment]
opravil jsem pár názvu proměnných a na jednom místě i indenty, proklikal http/s3 extraktor, za mě je to taky ready na merge, aby se mohlo pohnout dál

**Comment by ondrejhlavacek** [comment]
@kacurez je to ready, chceš si to mergnout? :-)

**Comment by kacurez** [comment]
ok mergnem to

**Pull request merged**